### DISCUSSION TITLE
Looking for way to test nexus resolvers/graphQL API with mocked Prisma client

### DISCUSSION THREAD

**Discussion by candidia** [graphql-nexus/nexus#1105] [Q&A]
The official docs only mention how to test nexus and prisma with a live server https://nexusjs.org/docs/getting-started/tutorial/chapter-testing-with-prisma/ but I'd like to test with a mocked Prisma client.

A possible workflow I have in mind is mocking the state of the prisma client db and then performing a query or mutation and testing the the state is updated appropriately. I'd also like to be able to execute the query as a user.

Is this not a common use case? Having trouble finding anything.

**Comment by ericyd** [comment]
What kinds of problems are you running into? It seems like a reasonable thing to do with Jest or some some other testing library that supports mocks.

That said, what do you mean by "testing the the state is updated appropriately"? To me this sounds like you want to assert that the underlying database was updated. In that case, shouldn't you just use the prisma client without mocking it?

**Reply by candidia** [reply]
Using the prisma client necessitates connecting to an actual db. Since I use Postgres in production, I can't use a different adapter because the types and features it supports are different. So just unsure how to do this without having to use a real postgres instance

**Reply by ericyd** [reply]
There is a massive body of writing online about how to test things that interact with a database, so I think the general question here is outside the scope of Nexus.

The simplest way to do this, and probably the way that is most common these days, is to run a local Postgres database that your tests can interact with. If you aren't already, I'd strongly recommend using a `.env` file or similar to store your DB URL, so you can easily configure it to your local test DB when running tests. This is what my team does and it works great. All major CI services allow you to spin up a Postgres container along side your tests for the same purpose.

There are several alternatives, but for the use case of integration testing an API, probably your most realistic alternative option is to mock the Prisma client, something like this:

```ts
jest.mock('prisma', () => ({
  Prisma: function () {
    // this is your mock class
    // here you can define any methods you want your mock Prisma client to return
  }
}))
```

I personally don't love the way Jest handles mocking classes exported from a library but you can read more in [their official docs](https://jestjs.io/docs/es6-class-mocks).

I hope this helps somewhat

**Reply by candidia** [reply]
Should have expanded on my reasoning. I want to avoid using Docker or a real DB instance because of how heavy they are. I feel like Prisma could add a mocked client to its library. I found a pretty good one https://www.npmjs.com/package/prisma-mock but it's not technically open source and lacks some parity with the prisma client like throwing errors when a record is not found.
